extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableChannelProvider.java

```diff
@@ -47,27 +49,65 @@
  */
 final class S3SeekableChannelProvider implements SeekableChannelsProvider {
 
+    private static final int MAX_KEYS_PER_BATCH = 1000;
+
+    private static final Logger log = LoggerFactory.getLogger(S3SeekableChannelProvider.class);
+
     /**
-     * We always allocate buffers of maximum allowed size for re-usability across reads with different fragment sizes.
-     * There can be a performance penalty though if the fragment size is much smaller than the maximum size.
+     * Cache of shared {@link S3AsyncClient} objects and request caches, keyed by the S3 instructions.
      */
-    private static final BufferPool BUFFER_POOL = new BufferPool(S3Instructions.MAX_FRAGMENT_SIZE);
+    private static final Map<S3Instructions, ClientData> SHARED_CLIENT_DATA = new ConcurrentHashMap<>();
 
-    private static final int MAX_KEYS_PER_BATCH = 1000;
+    private static class ClientData { // TODO better name
+        private final S3AsyncClient client;
+        private final S3RequestCache requestCache;
 
-    private static final Logger log = LoggerFactory.getLogger(S3SeekableChannelProvider.class);
+        ClientData(final S3AsyncClient client, final S3RequestCache requestCache) {
+            this.client = client;
+            this.requestCache = requestCache;
+        }
+    }
 
-    private final S3AsyncClient s3AsyncClient;
-    private final S3Instructions s3Instructions;
+    // Local references of the shared client and request cache
+    private final S3AsyncClient sharedAsyncClient;
+    private final S3RequestCache sharedCache;
 
-    S3SeekableChannelProvider(@NotNull final S3Instructions s3Instructions) {
-        // TODO(deephaven-core#5062): Add support for async client recovery and auto-close
-        // TODO(deephaven-core#5063): Add support for caching clients for re-use
-        this.s3AsyncClient = buildClient(s3Instructions);
-        this.s3Instructions = s3Instructions;
+    private final S3Instructions instructions;
+
+    S3SeekableChannelProvider(@NotNull final S3Instructions instructions) {
+        final ClientData clientData = SHARED_CLIENT_DATA.compute(instructions, (key, existingClientData) -> {
+            if (existingClientData == null) {
+                // No existing client, create a new one
+                final S3AsyncClient newClient = buildClient(instructions);
+                final S3RequestCache newCache = new ModuloBasedRequestCache(instructions.maxCacheSize());
+                return new ClientData(newClient, newCache);
+            } else {
+                // Client exists, reconnect if necessary
+                if (isConnectionOpen(existingClientData.client)) {
+                    return existingClientData;
+                }
+                final S3AsyncClient newClient = buildClient(instructions);
+                return new ClientData(newClient, existingClientData.requestCache);
+            }
+        });
```

Comment:
This seems suspicious. It looks like we are still creating a new client for every unique `s3Instructions`? It seems like we might want to key based on a subset of the S3Instructions fields; ultimately, so we could use the same client for different URIs against the same cloud provider.

I imagine we need the key to be at credentials, endpointOverride, and timeouts? I do question exactly what the meaning of some of our existing configurations mean in this new cache model. ie, maxCacheSize.

Reply:
There are quite a lot of instructions that go directly into building the shared client and cache.
- For building the client, we use `maxConcurrentRequests`, `connectionTimeout`, `readTimeout`, `regionName`, `credentialsProvider`, and `endpointOverride`.
- For the cache, we use `maxCacheSize` for building a shared cache, and `fragmentSize` to find the fragment index which is used index the cache.

Other than these, only `readAheadCount` is left. That is why I kept all to avoid missing any parameters which get added in the future and get left out.

And regrading maxCacheSize, in this new model, it is controlling the number of cached elements per context object as well as the maximum number of cached elements in the shared cache.